### ISSUE TITLE
OSC restored on wrong position in fs with cocoa-cb

### ISSUE THREAD

**Issue by AirPort** [mpv-player/mpv#5525]
### mpv version and platform

mpv git
macOS 10.13.3

### Reproduction steps

It's basically a minor nitpick, but I thought I might as well report it: open a video file with a lower height than your screen (eg. a common 16:9 movie on a 16:10 screen). Exit in fullscreen with Shift+Q to save position, then reopen the file again.

### Expected behavior

The OSC (bottombar) is displayed all the way to the bottom of the screen.

### Actual behavior

The OSC is displayed on the bottom _of the video_, above the black bar. Going on and off fs "restore" the position to the expected one.

### Log file

Screenshots are probably more useful: 
[Regular position](https://0x0.st/sTZY.png)
[Restored position](https://0x0.st/sTZg.png)

**Comment by Akemi** [comment]
weird there is an extra setFrame function call after the animation that comes out of nowhere.

**Comment by Akemi** [comment]
it also just happens with `mpv FILE --fs`, a bit faster to reproduce like that.

**Comment by Akemi** [comment]
should be fixed with #5529.

**Comment by AirPort** [comment]
Kind of the opposite, actually: now the osd seems to be always drawn in the "wrong" position, inside the video.

**Comment by Akemi** [comment]
that's weird i removed that commit from the PR for now.

**Comment by Akemi** [comment]
does this work?
```
diff --git a/video/out/cocoa-cb/window.swift b/video/out/cocoa-cb/window.swift
index c93537c..da3e319 100644
--- a/video/out/cocoa-cb/window.swift
+++ b/video/out/cocoa-cb/window.swift
@@ -167,10 +167,11 @@ class Window: NSWindow, NSWindowDelegate {
     }
 
     func windowDidEnterFullScreen(_ notification: Notification) {
+        let newFrame = targetScreen!.frame
         isInFullscreen = true
         cocoaCB.flagEvents(VO_EVENT_FULLSCREEN_STATE)
         cocoaCB.updateCusorVisibility()
-        endAnimation()
+        endAnimation(newFrame)
     }
 
     func windowDidExitFullScreen(_ notification: Notification) {
@@ -182,16 +183,23 @@ class Window: NSWindow, NSWindowDelegate {
     func windowDidFailToEnterFullScreen(_ window: NSWindow) {
         let newFrame = calculateWindowPosition(for: targetScreen!, withoutBounds: targetScreen == screen)
         setFrame(newFrame, display: true)
-        endAnimation()
+        endAnimation(newFrame)
     }
 
     func windowDidFailToExitFullScreen(_ window: NSWindow) {
         let newFrame = targetScreen!.frame
         setFrame(newFrame, display: true)
-        endAnimation()
+        endAnimation(newFrame)
     }
 
-    func endAnimation() {
+    func endAnimation(_ rect: NSRect = NSZeroRect) {
+        if !NSIsEmptyRect(rect) {
+            Swift.print("endAnimation")
+            NSAnimationContext.runAnimationGroup({ (context) -> Void in
+                context.duration = 0.01
+                self.animator().setFrame(frame, display: true)
+            }, completionHandler: nil )
+        }
         isAnimating = false
         cocoaCB.isShuttingDown = false
     }
```

**Comment by AirPort** [comment]
Nope, sorry, it's same as before, with OSC is displayed only at the bottom of the video. 
It also introduces a new issue compared to current master: if you restore a video in fullscreen, it appears unscaled, with black bars around all four borders (I checked, this was also true for the previous fix).

**Comment by Akemi** [comment]
how about this? it isn't really a fix, more like a workaround. also could you by chance make a screen recording of the restoring with the unscaled output? i was wondering how the animation looks like.
```
diff --git a/video/out/cocoa_cb_common.swift b/video/out/cocoa_cb_common.swift
index 0ce0642..cd43195 100644
--- a/video/out/cocoa_cb_common.swift
+++ b/video/out/cocoa_cb_common.swift
@@ -114,7 +114,9 @@ class CocoaCB: NSObject {
         layer.setVideo(true)
 
         if self.mpv.getBoolProperty("fullscreen") {
-            window.toggleFullScreen(nil)
+            DispatchQueue.main.async {
+                self.window.toggleFullScreen(nil)
+            }
         } else {
             window.isMovableByWindowBackground = true
         }
```

**Comment by AirPort** [comment]
Yep, this workaround works, the OSC is always drawn at the bottom of the screen, and the video is restored in fs with the right dimensions.
[Here's](https://0x0.st/sTaw.MOV) a recording of the restoring with unscaled output from the previous fix. Sorry for the shitty telephone quality, but I didn't want to potentially mess with the results using QuickTime; if it's too garbage to be useful tell me and I'll try to make a better one.

**Comment by Akemi** [comment]
it would be nice if you could apply [this](https://0x0.st/sTQz.diff) patch to master and give me your terminal output from the issue.

**Comment by AirPort** [comment]
Sorry, I'm a bit late. [Here's](https://0x0.st/sTei.txt) the log with the patch applied to current master (I also applied the dependencies patch from the other PR to avoid hassles during build, I hope it doesn't change anything).

**Comment by Akemi** [comment]
@AirPort thx that was very helpful. i could identify several problems that only manifest themselves in very specific situations. it depends on timing because there is quite a few async calls/events and some events being called a bit erratic resulting in the async drawing path to switch on/off several times in a under 1 sec. it will probably take a bit longer fixing them properly.

on a side note, the first patch actually fixed one problem but made another one more apparent in your case.

**Comment by Akemi** [comment]
should be fixed with #5563.

**Comment by AirPort** [comment]
Yep, as far as I can see it's indeed fixed.
